frontend/src/App.js

```diff
@@ -1,27 +1,51 @@
 import React, { useState, useEffect } from 'react';
-import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
+import {
+  BrowserRouter as Router,
+  Route,
+  Switch,
+  useLocation,
+  useHistory,
+} from 'react-router-dom';
+import jwt_decode from 'jwt-decode';
 import { ThemeProvider } from '@material-ui/core/styles';
 import CssBaseline from '@material-ui/core/CssBaseline';
+import Container from '@material-ui/core/Container';
+import Box from '@material-ui/core/Box';
 import theme from './styles/theme';
-import jwt_decode from 'jwt-decode';
 
 import Hypertube from './components/gallery/Index';
 import Landing from './components/layout/Landing';
 import Navbar from './components/layout/Navbar';
 import setAuthToken from './utils/setAuthToken';
-
-import Box from '@material-ui/core/Box';
-import { Container } from 'react-bootstrap';
+import authService from './services/auth';
 
 function App() {
   const [user, setUser] = useState({ userId: '', lang: '' });
+  const location = useLocation();
+  const history = useHistory();
 
   useEffect(() => {
-    setAuthToken(localStorage.getItem('token'));
-    if (localStorage.getItem('token')) {
-      const decoded = jwt_decode(localStorage.getItem('token'));
+    const setUserFromToken = (token) => {
+      const decoded = jwt_decode(token);
       setUser({ userId: decoded.id, lang: decoded.lang });
-    }
+			setAuthToken(token);
+    };
+
+    const checkToken = async () => {
+      try {
+        const { token } = await authService.getToken();
+        setUserFromToken(token);
+				history.push('/');
+      } catch (e) {
+        console.log(e);
+      }
+    };
+
+    if (location.search === '?auth=token') {
```

Comment:
With this approach, we have next issue. 
I logged in as some user (with fake email), then in incognito mode logged in with gmail account. 
In normal mode as some user, I added in url : ?auth=token
And after that I was logged in not as usual user with fake email, but as user with gmail account.

Reply:
I was thinking of changing this to approach 1: store token in an array or object and send the index/key to front: instead of `auth=token` `auth=<index/key>`
Would that be safe enough for our needs? I could use uuid as a key, so it would be very unlikely that user would guess it at least manually.

Reply:
Yes, this would be better, I think (using uuid as a key, and setting to null, if fetched)
Also maybe in App.js first we should check if localStorage has token, and if not then check if location is auth=<index/key>?
So if user logged in and suddenly put ?auth=....., then function checkToken will not be called

Reply:
This is now done 🙂
- tokens are stored to an object with uuid keys and available only once
- getToken is only called when there's no token in localStorage and `auth=<key>`